docs/synapse/userguides/storm_adv_functions.rstorm

```diff
@@ -0,0 +1,783 @@
+
+Or:
+
+::
+  
+  $asn=:asn
+  $myFunction($asn)
+
+Alternatively, you can pass the entire ``$node`` object to the function and use the ``yield`` keyword
+within the function to yield the node into the function's pipeline:
+
+::
+  
+  //Declare function
+  function $myFunction(inboundNode) {
+  
+      yield $inboundNode
+      <do stuff>
+      return()
+  }
+  
+  //Invoke function
+  $myFunction($node)
+
+If another function yields the node(s) you want to operate on, that function can be used as input to a
+second function. A simple example:
+
+**Function 1 (node yielder):**
+
+::
+  
+  function getIPs() {
+      //Lift 10 IPv4 addresses
+      inet:ipv4 | limit 10
+  }
+
+**Function 2 (callable function):**
+
+::
+  
+  //Takes a generator object as input
+  function counter(genr) {
+      
+      //Yield the generator content into the function pipeline
+      yield $genr
+      
+      //Print the human-readable representation of each node
+      $lib.print($node.repr())
+      
+      //Return the output of the function
+      fini { return() }
+  }
+
+**Function 2 invoked with Function 1 as input:**
+
+::
+  
+  $counter($getIPs())
+
+When executed, the function produces the following output. Note that the ``$counter()`` function simply prints
+the nodes' human-readable representation (``$node.repr()``) as an example; it does not return or yield the
+``inet:ipv4`` nodes:
+
+.. storm-pre:: [ inet:ipv4=1.1.1.1 inet:ipv4=2.2.2.2 inet:ipv4=3.3.3.3 inet:ipv4=4.4.4.4 inet:ipv4=5.5.5.5 inet:ipv4=6.6.6.6 inet:ipv4=7.7.7.7 inet:ipv4=8.8.8.8 inet:ipv4=9.9.9.9 inet:ipv4=10.10.10.10 ]
+
+.. storm-multiline:: FUNCTIONS="\nfunction getIPs() { \n    inet:ipv4 | limit 10 \n} \n\nfunction counter(genr) {\n    yield $genr \n    $lib.print($node.repr()) \n    fini { return() } \n} \n\n$counter($getIPs())"
+
+.. storm-cli:: MULTILINE=FUNCTIONS
+
+
+.. _storm-func-runtsafe:
+
+Runtsafe vs. Non-Runtsafe Functions
+***********************************
+
+Just as variables may be runtime-safe (`runtsafe`_) or non-runtime-safe (`non-runtsafe`_), functions can be
+invoked in a runtsafe manner (or not) based on the parameters passed to the function.
+
+If a function is invoked with a runtsafe (typically static) value, the function is considered runtsafe. A
+function that takes an Autonomous System (AS) number as input and is passed a static AS number as a parameter
+is invoked in a **runtsafe** manner:
+
+::
+  
+  $myFunction(9009)
+
+Or:
+
+::
+  
+  $asn=9009
+  myFunction($asn)
+
+If the same function is invoked with a per-node, non-runtsafe value or values, the function is considered
+**non-runtsafe,** such as the example above where the invoking pipeline contains ``inet:ipv4`` nodes and the
+function is invoked with the value of each node's ``:asn`` property:
+
+::
+  
+  $myFunction(:asn)
+
+Or:
+
+::
+  
+  $asn=:asn
+  $myFunction($asn)
+
+.. TIP::
+  
+  Keep in mind that functions execute when they are invoked, regardless of whether there are nodes in
```

Comment:
Seems like this line needs a little adjusting too, "execute when they are invoked, regardless of whether there are nodes in the invoking pipeline." conflicts with "A non-runtsafe function (i.e., that is dependent on a per-node value) will **not** execute when invoked if there are no nodes in the invoking pipeline"

Reply:
revised :upside_down_face: